### PR TITLE
feat(mls): update keying material every 24 hrs (AR-2282)

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Check the limited duration(here 24hrs) to update keying material. and if any of updates failed we don't update the update time.

### Issues

To do more optimisation.

### Causes (Optional)
To do more optimisation.

### Solutions

Add a timestamp key to check the period.

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/TimestampKeyRepository.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.wrapStorageRequest
+import com.wire.kalium.persistence.dao.MetadataDAO
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import kotlin.time.Duration
+
+interface TimestampKeyRepository {
+    suspend fun isPassed(key: String, duration: Duration): Either<StorageFailure, Boolean>
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/TimestampKeyRepository.kt:14]:
```suggestion
    suspend fun hasPassed(key: String, duration: Duration): Either<StorageFailure, Boolean>
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/KeyingMaterialsManager.kt [hunk 2]
```diff
@@ -14,12 +17,8 @@ import kotlinx.coroutines.launch
 import kotlin.time.Duration.Companion.hours
 
 // The duration in hours after which we should re-check key package count.
-/**
- * TODO: ATM we don't need that, but later we can optimise the manager by running it only every 24hours
- *   If we impl the 24hours check, then we need to check if all the update call succeed,
- *    other than that we should not consider the update as succeeded.
- */
 internal val KEYING_MATERIAL_CHECK_DURATION = 24.hours
+internal const val LAST_KEYING_MATERIAL_UPDATE_CHECK = "LAST_KEYING_MATERIAL_UPDATE_CHECK"
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/KeyingMaterialsManager.kt:21]: Would it make sense to move this into a `TimestampKeys` enum?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt [hunk 3]
```diff
@@ -24,9 +26,15 @@ class UpdateKeyingMaterialsUseCaseImpl(
...
+                    mlsConversationRepository.updateKeyingMaterial(groupId).onFailure {
+                        updatesSucceeded = false
+                    }
                 }
+                if (updatesSucceeded)
+                    Either.Right(Unit)
+                else Either.Left(Unit)
             }.fold(
                 { UpdateKeyingMaterialsResult.Failure(it) },
                 { UpdateKeyingMaterialsResult.Success }
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt:40]: Can we simplify this by something like this:

```
groups.map { mlsConversationRepository.updateKeyingMaterial(it) }
  .foldToEitherWhileRight(Unit) { value, _ ->
    value
  }.fold(
    { UpdateKeyingMaterialsResult.Failure(it) },
    { UpdateKeyingMaterialsResult.Success }
  )
```

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt (hunk 3 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt:40]: Clean advise! But one question, in that case we stop the whole loop, am I right? My idea is updating as much as we can in each call.

Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt:40]: Unless Either map is lazy will still update all conversations `foldToEitherWhileRight` will stop on the first error when mapping over a list of results.

**Review by typfel — APPROVED** [review]
Nice!

**Pull request merged**